mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,416 @@
+                let right = &mmap[0..split];
+                let mut bytes = Vec::from(left);
+                bytes.append(&mut Vec::from(right));
+                data = binary_deserialize::<Data>(&bytes)?;
+            }
+            validate(&block, &data)?;
+            let something = binary_deserialize::<Vec<u8>>(data.inner())?;
+            vdata.push_back((index, something));
+            read_index += block_size + data_size;
+        }
+
+        self.pointers.read.store(
+            (read_index + total_size) % self.max_file_size,
+            Ordering::SeqCst,
+        );
+
+        Poll::Ready(Ok(vdata))
+    }
+}
+
+struct RingBufferRemoveFuture {
+    pointers: Arc<FilePointers>,
+    flush_options: FlushOptions,
+    mmap: Arc<Mutex<MmapMut>>,
+    flush_state: Arc<FlushState>,
+    key: usize,
+    max_file_size: usize,
+}
+
+impl Future for RingBufferRemoveFuture {
+    type Output = StorageResult<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let timer = Instant::now();
+        let write_index = self.pointers.write.load(Ordering::SeqCst);
+        let block_size_result = binary_serialize_size(&BlockHeaderWithHash::new(0, 0, 0));
+
+        if block_size_result.is_err() {
+            return Poll::Ready(Err(RingBufferError::Serialization(
+                block_size_result.unwrap_err(),
+            )
+            .into()));
+        }
+
+        let block_size = block_size_result.unwrap();
+
+        if self.key >= write_index {
+            return Poll::Pending;
+        }
+        cx.waker().wake_by_ref();
+
+        let start = self.key;
+        let end = start + block_size;
+        if end > self.max_file_size {
+            return Poll::Ready(Err(RingBufferError::WrapAround.into()));
+        }
+        let bytes = mmap_read(self.mmap.clone(), start, end);
+
+        let mut block = binary_deserialize::<BlockHeaderWithHash>(&bytes)?;
+
+        let inner_block = block.inner_mut();
+        inner_block.set_should_not_overwrite(false);
+
+        let bytes = binary_serialize(&block)?;
+
+        let should_flush = should_flush(self.flush_options, &self.flush_state);
+        mmap_write(self.mmap.clone(), start, end, &bytes, should_flush)?;
+
+        self.flush_state
+            .update(0, 0, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl Storage<usize, Vec<u8>> for RingBuffer {
+    fn insert_future<'a>(
+        &self,
+        value: &Vec<u8>,
+    ) -> Pin<Box<dyn Future<Output = StorageResult<()>> + Send + 'a>> {
+        let future = RingBufferWriteFuture {
+            pointers: self.pointers.clone(),
+            flush_options: self.flush_options,
+            mmap: self.inner.clone(),
+            flush_state: self.flush_state.clone(),
+            value: value.clone(),
```

Comment:
can we move the buffer instead of clone?

Reply:
Yeah cloning the data is more than I want, let me update this :)